File: src/Data/String/CodeUnits.erl
-module(data_string_codeUnits@foreign).
-export([fromCharArray/1, toCharArray/1, singleton/1, '_charAt'/4,  '_toChar'/3,length/1, '_indexOf'/4, '_indexOfStartingAt'/5,'_lastIndexOf'/4,'_lastIndexOfStartingAt'/5,take/2, drop/2, countPrefix/2, '_slice'/3, splitAt/2]).

fromCharArray(A) -> case unicode:characters_to_binary(array:to_list(A)) of
  {error, S, _} -> S; % In case of hitting bad character, return initial portion!
  S -> S
end.

toCharArray(S) -> array:from_list(unicode:characters_to_list(S)).

singleton(C) -> <<C>>.

'_charAt'(Just,Nothing,I,S) ->
  case byte_size(S) of
    N when I >= 0, I < N -> Just(binary:at(S, I));
    _ -> Nothing
  end.

'_toChar'(Just,_,<<C>>) -> Just(C);
'_toChar'(_,Nothing,_) -> Nothing.

length(S) -> byte_size(S).

'_indexOf'(Just,_,<<>>,_) -> Just(0);
'_indexOf'(Just,Nothing,X,S) ->
  case binary:match(S, X) of
    nomatch -> Nothing;
    { Start, _Length } -> Just(Start)
  end.
'_indexOfStartingAt'(_Just,Nothing,_X,StartAt,S) when StartAt < 0; StartAt > byte_size(S) -> Nothing;
'_indexOfStartingAt'(Just,_Nothing,<<>>,StartAt,_S) -> Just(StartAt);
'_indexOfStartingAt'(Just,Nothing,X,StartAt,S) ->
  case binary:match(S, X, [{scope, {StartAt, byte_size(S)-StartAt}}]) of
    nomatch -> Nothing;
    { Start, _Length } -> Just(Start)
  end.

'_lastIndexOf'(Just,_,<<>>,S) -> Just(byte_size(S));
'_lastIndexOf'(Just,Nothing,X,S) ->
  % there is no binary reverse match, and matches returns only nonoverlapping so would not always include the last match
  case string:find(S, X, trailing) of
    nomatch -> Nothing;
    Substr -> Just(byte_size(S) - byte_size(Substr))
  end.
'_lastIndexOfStartingAt'(Just,Nothing,X,StartAt,S) ->
  L = byte_size(S),
  S0 = min(L, max(0, StartAt)),
  Part = binary:part(S, {0, S0+byte_size(X)}),
  case '_lastIndexOf'(Just,Nothing,X,Part) of
    {just,N} -> Just(N);
    _ -> Nothing
  end.

take(N,_) when N < 0 -> <<>>;
take(N,S) when N > byte_size(S) -> S;
take(N,S) -> binary:part(S, {0, N}).

drop(N,S) ->
  M = max(0, min(N, byte_size(S))),
  binary:part(S, {M, byte_size(S)-M}).

countPrefixImpl(_, S, I) when I >= byte_size(S) -> I;
countPrefixImpl(P, S, I) -> 
    case P(binary:at(S, I)) of
        true -> countPrefixImpl(P, S, I+1);
        false -> I
    end.

countPrefix(P, S) -> countPrefixImpl(P, S, 0).

'_slice'(B, E, S) -> 
  L = byte_size(S),
  BB = case B < 0 of
        true -> L + B;
        false -> B
       end,
  EE = case E < 0 of
        true -> L + E;
        false -> E
       end,
  binary:part(S, BB, (EE-BB)).

splitAt(I0, S) ->
    I = min(max(0, I0), erlang:byte_size(S)),
    Start = binary:part(S, 0, I),
    End = binary:part(S, I, erlang:byte_size(S)-I),
    #{before => Start, 'after' => End}.

File: src/Data/String/CodeUnits.js
"use strict";

export var fromCharArray = function (a) {
  return a.join("");
};

export var toCharArray = function (s) {
  return s.split("");
};

export var singleton = function (c) {
  return c;
};

export var _charAt = function (just) {
  return function (nothing) {
    return function (i) {
      return function (s) {
        return i >= 0 && i < s.length ? just(s.charAt(i)) : nothing;
      };
    };
  };
};

export var _toChar = function (just) {
  return function (nothing) {
    return function (s) {
      return s.length === 1 ? just(s) : nothing;
    };
  };
};

export var length = function (s) {
  return s.length;
};

export var countPrefix = function (p) {
  return function (s) {
    var i = 0;
    while (i < s.length && p(s.charAt(i))) i++;
    return i;
  };
};

export var _indexOf = function (just) {
  return function (nothing) {
    return function (x) {
      return function (s) {
        var i = s.indexOf(x);
        return i === -1 ? nothing : just(i);
      };
    };
  };
};

export var _indexOfStartingAt = function (just) {
  return function (nothing) {
    return function (x) {
      return function (startAt) {
        return function (s) {
          if (startAt < 0 || startAt > s.length) return nothing;
          var i = s.indexOf(x, startAt);
          return i === -1 ? nothing : just(i);
        };
      };
    };
  };
};

export var _lastIndexOf = function (just) {
  return function (nothing) {
    return function (x) {
      return function (s) {
        var i = s.lastIndexOf(x);
        return i === -1 ? nothing : just(i);
      };
    };
  };
};

export var _lastIndexOfStartingAt = function (just) {
  return function (nothing) {
    return function (x) {
      return function (startAt) {
        return function (s) {
          var i = s.lastIndexOf(x, startAt);
          return i === -1 ? nothing : just(i);
        };
      };
    };
  };
};

export var take = function (n) {
  return function (s) {
    return s.substr(0, n);
  };
};

export var drop = function (n) {
  return function (s) {
    return s.substring(n);
  };
};

export var _slice = function (b) {
  return function (e) {
    return function (s) {
      return s.slice(b,e);
    };
  };
};

export var splitAt = function (i) {
  return function (s) {
    return { before: s.substring(0, i), after: s.substring(i) };
  };
};
